app/src/main/java/org/mozilla/rocket/preference/SharedPreferenceLiveData.kt

```diff
@@ -0,0 +1,74 @@
+package org.mozilla.rocket.preference
+
+import android.arch.lifecycle.LiveData
+import android.content.SharedPreferences
+
+abstract class SharedPreferenceLiveData<T>(
+    val sharedPrefs: SharedPreferences,
+    val key: String,
+    private val defValue: T
+) : LiveData<T>() {
+
+    private val preferenceChangeListener = SharedPreferences.OnSharedPreferenceChangeListener { _, key ->
+        if (key == this.key) {
+            value = getValueFromPreferences(key, defValue)
+        }
+    }
+
+    abstract fun getValueFromPreferences(key: String, defValue: T): T
+
+    override fun onActive() {
+        super.onActive()
+        value = getValueFromPreferences(key, defValue)
+        sharedPrefs.registerOnSharedPreferenceChangeListener(preferenceChangeListener)
+    }
+
+    override fun onInactive() {
+        sharedPrefs.unregisterOnSharedPreferenceChangeListener(preferenceChangeListener)
+        super.onInactive()
+    }
+
+    fun forceNotify() {
+        value = value
+    }
+}
+
+class SharedPreferenceIntLiveData(sharedPrefs: SharedPreferences, key: String, defValue: Int) :
+        SharedPreferenceLiveData<Int>(sharedPrefs, key, defValue) {
+    override fun getValueFromPreferences(key: String, defValue: Int): Int = sharedPrefs.getInt(key, defValue)
+}
+
+class SharedPreferenceStringLiveData(sharedPrefs: SharedPreferences, key: String, defValue: String) :
+        SharedPreferenceLiveData<String>(sharedPrefs, key, defValue) {
+    override fun getValueFromPreferences(key: String, defValue: String): String = requireNotNull(sharedPrefs.getString(key, defValue))
```

Comment:
why do we need to requireNotNull here? If the key doesn't exist, the defValue, which is a non-null type variable, will be returned instead. It seems the requreNotNull here is unnecessary. This is not a merge blocker, just want to know if there's any special reason